File: benchmark_client/scripts/cpp/Fractals.cpp
#include <chrono>
#include <cmath>
#include <cstdio>
#include <vector>

namespace {

constexpr int W = 4000;
constexpr int H = 4000;
constexpr int MAX_ITER = 1000;

inline void mandelbrot_row(int py, int *out_row) {
    const float ymin = -1.5f;
    const float ymax = 1.5f;
    const float xmin = -2.0f;
    const float xmax = 1.0f;
    const float c_im = ymin + (float)py / (float)H * (ymax - ymin);

    for (int px = 0; px < W; ++px) {
        const float c_re = xmin + (float)px / (float)W * (xmax - xmin);
        float z_re = 0.0f;
        float z_im = 0.0f;
        int iter = 0;
        for (iter = 0; iter < MAX_ITER; ++iter) {
            const float z_re2 = z_re * z_re;
            const float z_im2 = z_im * z_im;
            if (z_re2 + z_im2 > 4.0f) {
                break;
            }
            z_im = 2.0f * z_re * z_im + c_im;
            z_re = z_re2 - z_im2 + c_re;
        }
        out_row[px] = iter;
    }
}

inline void julia_row(int py, int *out_row) {
    const float xmin = -1.6f;
    const float xmax = 1.6f;
    const float ymin = -1.6f;
    const float ymax = 1.6f;
    const float c_re_k = -0.7f;
    const float c_im_k = 0.27015f;
    const float z_im0 = ymin + (float)py / (float)H * (ymax - ymin);

    for (int px = 0; px < W; ++px) {
        float z_re = xmin + (float)px / (float)W * (xmax - xmin);
        float z_im = z_im0;
        int iter = 0;
        for (iter = 0; iter < MAX_ITER; ++iter) {
            const float z_re2 = z_re * z_re;
            const float z_im2 = z_im * z_im;
            if (z_re2 + z_im2 > 4.0f) {
                break;
            }
            z_im = 2.0f * z_re * z_im + c_im_k;
            z_re = z_re2 - z_im2 + c_re_k;
        }
        out_row[px] = iter;
    }
}

inline void burning_ship_row(int py, int *out_row) {
    const float xmin = -2.2f;
    const float xmax = 1.2f;
    const float ymin = -2.0f;
    const float ymax = 1.0f;
    const float c_im = ymin + (float)py / (float)H * (ymax - ymin);

    for (int px = 0; px < W; ++px) {
        const float c_re = xmin + (float)px / (float)W * (xmax - xmin);
        float z_re = 0.0f;
        float z_im = 0.0f;
        int iter = 0;
        for (iter = 0; iter < MAX_ITER; ++iter) {
            const float z_re2 = z_re * z_re;
            const float z_im2 = z_im * z_im;
            if (z_re2 + z_im2 > 4.0f) {
                break;
            }
            const float next_im = 2.0f * std::fabs(z_re) * std::fabs(z_im) + c_im;
            const float next_re = z_re2 - z_im2 + c_re;
            z_re = next_re;
            z_im = next_im;
        }
        out_row[px] = iter;
    }
}

void run_mandelbrot(std::vector<int> &buf) {
#pragma omp parallel for schedule(dynamic, 8)
    for (int py = 0; py < H; ++py) {
        mandelbrot_row(py, &buf[static_cast<std::size_t>(py) * W]);
    }
}

void run_julia(std::vector<int> &buf) {
#pragma omp parallel for schedule(dynamic, 8)
    for (int py = 0; py < H; ++py) {
        julia_row(py, &buf[static_cast<std::size_t>(py) * W]);
    }
}

void run_burning_ship(std::vector<int> &buf) {
#pragma omp parallel for schedule(dynamic, 8)
    for (int py = 0; py < H; ++py) {
        burning_ship_row(py, &buf[static_cast<std::size_t>(py) * W]);
    }
}

}  // namespace

int main() {
    std::vector<int> buf(static_cast<std::size_t>(W) * static_cast<std::size_t>(H));

    const auto t0 = std::chrono::high_resolution_clock::now();
    run_mandelbrot(buf);
    run_julia(buf);
    run_burning_ship(buf);
    const auto t1 = std::chrono::high_resolution_clock::now();
    const double elapsed_ms =
        std::chrono::duration<double, std::milli>(t1 - t0).count();

    (void)buf[0];
    std::printf("{\"execution_time_ms\": %.6f}\n", elapsed_ms);
    return 0;
}

File: benchmark_client/scripts/go/Fractals.go
package main

import (
	"encoding/json"
	"fmt"
	"math"
	"os"
	"sync"
	"time"
)

const (
	width    = 4000
	height   = 4000
	maxIter  = 1000
	rowBatch = 100
)

func mandelbrotRows(y0, y1 int, buf []int32) {
	const (
		xmin, xmax = -2.0, 1.0
		ymin, ymax = -1.5, 1.5
	)
	for py := y0; py < y1; py++ {
		cim := ymin + float64(py)/float64(height)*(ymax-ymin)
		rowOff := py * width
		for px := 0; px < width; px++ {
			cre := xmin + float64(px)/float64(width)*(xmax-xmin)
			zre, zim := 0.0, 0.0
			iter := 0
			for iter = 0; iter < maxIter; iter++ {
				zre2 := zre * zre
				zim2 := zim * zim
				if zre2+zim2 > 4.0 {
					break
				}
				zim = 2.0*zre*zim + cim
				zre = zre2 - zim2 + cre
			}
			buf[rowOff+px] = int32(iter)
		}
	}
}

func juliaRows(y0, y1 int, buf []int32) {
	const (
		xmin, xmax = -1.6, 1.6
		ymin, ymax = -1.6, 1.6
		cRe, cIm   = -0.7, 0.27015
	)
	for py := y0; py < y1; py++ {
		zim0 := ymin + float64(py)/float64(height)*(ymax-ymin)
		rowOff := py * width
		for px := 0; px < width; px++ {
			zre := xmin + float64(px)/float64(width)*(xmax-xmin)
			zim := zim0
			iter := 0
			for iter = 0; iter < maxIter; iter++ {
				zre2 := zre * zre
				zim2 := zim * zim
				if zre2+zim2 > 4.0 {
					break
				}
				zim = 2.0*zre*zim + cIm
				zre = zre2 - zim2 + cRe
			}
			buf[rowOff+px] = int32(iter)
		}
	}
}

func burningShipRows(y0, y1 int, buf []int32) {
	const (
		xmin, xmax = -2.2, 1.2
		ymin, ymax = -2.0, 1.0
	)
	for py := y0; py < y1; py++ {
		cim := ymin + float64(py)/float64(height)*(ymax-ymin)
		rowOff := py * width
		for px := 0; px < width; px++ {
			cre := xmin + float64(px)/float64(width)*(xmax-xmin)
			zre, zim := 0.0, 0.0
			iter := 0
			for iter = 0; iter < maxIter; iter++ {
				zre2 := zre * zre
				zim2 := zim * zim
				if zre2+zim2 > 4.0 {
					break
				}
				nextIm := 2.0*math.Abs(zre)*math.Abs(zim) + cim
				nextRe := zre2 - zim2 + cre
				zre, zim = nextRe, nextIm
			}
			buf[rowOff+px] = int32(iter)
		}
	}
}

func parallelByRows(compute func(y0, y1 int, buf []int32), buf []int32) {
	var wg sync.WaitGroup
	for y0 := 0; y0 < height; y0 += rowBatch {
		y1 := y0 + rowBatch
		if y1 > height {
			y1 = height
		}
		wg.Add(1)
		go func(a, b int) {
			defer wg.Done()
			compute(a, b, buf)
		}(y0, y1)
	}
	wg.Wait()
}

func main() {
	buf := make([]int32, width*height)

	start := time.Now()
	parallelByRows(mandelbrotRows, buf)
	parallelByRows(juliaRows, buf)
	parallelByRows(burningShipRows, buf)
	elapsedMs := float64(time.Since(start).Nanoseconds()) / 1e6

	_ = buf[0]

	out, err := json.Marshal(map[string]float64{"execution_time_ms": elapsedMs})
	if err != nil {
		os.Exit(1)
	}
	fmt.Println(string(out))
}
